typescript-generator-core/src/main/java/cz/habarta/typescript/generator/ext/BeanPropertyPathExtension.java

```diff
@@ -0,0 +1,136 @@
+            writer.writeIndentedLine(fieldsClassLine.replace("    ", settings.indentString));
+        }
+    }
+
+    /**
+     * Emits a bean and its parent beans before if needed.
+     * Returns the list of beans that were emitted.
+     */
+    private static Set<TsBeanModel> writeBeanAndParentsFieldSpecs(
+        Writer writer, Settings settings, TsModel model, Set<TsBeanModel> emittedSoFar, TsBeanModel bean) {
+        if (emittedSoFar.contains(bean)) {
+            return new HashSet<>();
+        }
+        final TsBeanModel parentBean = getBeanModelByType(model, bean.getParent());
+        final Set<TsBeanModel> emittedBeans = parentBean != null
+            ? writeBeanAndParentsFieldSpecs(writer, settings, model, emittedSoFar, parentBean)
+            : new HashSet<TsBeanModel>();
+        final String parentClassName = parentBean != null
+            ? getBeanModelClassName(parentBean) + "Fields"
+            : "Fields";
+        writer.writeIndentedLine("");
+        writer.writeIndentedLine(
+            "class " + getBeanModelClassName(bean) + "Fields extends " + parentClassName + " {");
+        writer.writeIndentedLine(
+            settings.indentString + "constructor(parent?: Fields, name?: string) { super(parent, name); }");
+        for (TsPropertyModel property : bean.getProperties()) {
+            writeBeanProperty(writer, settings, model, bean, property);
+        }
+        writer.writeIndentedLine("}");
+
+        emittedBeans.add(bean);
+        return emittedBeans;
+    }
+
+    private static TsBeanModel getBeanModelByType(TsModel model, TsType type) {
+        for (TsBeanModel curBean : model.getBeans()) {
+            if (curBean.getName().equals(type)) {
+                return curBean;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * return a class name formatted for rendering in code
+     * as part of another class name (so, for generics, strip
+     * the type arguments)
+     */
+    private static String getBeanModelClassName(TsBeanModel bean) {
+        return bean.getName() instanceof GenericReferenceType
+            ? ((GenericReferenceType)bean.getName()).symbol.toString()
+            : bean.getName().toString();
+    }
+
+    private static void writeBeanProperty(
+        Writer writer, Settings settings, TsModel model, TsBeanModel bean,
+        TsPropertyModel property) {
+        TsBeanModel fieldBeanModel = getBeanModelByType(model, property.getTsType());
+        String fieldClassName = fieldBeanModel != null ? getBeanModelClassName(fieldBeanModel) : "";
+        writer.writeIndentedLine(
+            settings.indentString + property.getName() + " = new " + fieldClassName + "Fields(this, \"" + property.getName() + "\");");
+    }
+
+    private static void createBeanFieldConstant(Writer writer, TsBeanModel bean) {
+        writer.writeIndentedLine("export const " + getBeanModelClassName(bean) + " = new " + getBeanModelClassName(bean) + "Fields();");
```

Comment:
Done. I thought you need `export` when generating .ts as opposed to .d.ts -- for us here without export, the typescript code doesn't build.